templates/templates.go

```diff
@@ -0,0 +1,418 @@
+			hours := int64(d.Hours()) % 24
+			minutes := int64(d.Minutes()) % 60
+			seconds := int64(d.Seconds()) % 60
+
+			segments := []struct {
+				name  string
+				value int64
+			}{
+				{"Day", days},
+				{"Hour", hours},
+				{"Min", minutes},
+				{"Sec", seconds},
+			}
+
+			var parts []string
+
+			for _, s := range segments {
+				if s.value == 0 {
+					continue
+				}
+				plural := ""
+				if s.value != 1 {
+					plural = "s"
+				}
+
+				parts = append(parts, fmt.Sprintf("%d %s%s", s.value, s.name, plural))
+			}
+			return strings.Join(parts, " ")
+		},
+		"formatTime": func(fmt string, t time.Time) string {
+			return t.Format(fmt)
+		},
+		"now": func() time.Time {
+			return time.Now()
+		},
+		"thisYear": func() int {
+			return time.Now().Year()
+		},
+		"add": func(a, b int) int {
+			return a + b
+		},
+		"inc": func(a int) int {
+			return a + 1
+		},
+		"dec": func(a int) int {
+			return a - 1
+		},
+		"even": func(a int) bool {
+			return a%2 == 0
+		},
+		"len": func(a string) int {
+			return len(a)
+		},
+		"checkPermission": func(perms []permission.Permission, p string) bool {
+			return t.permissionsParser(perms, p)
+		},
+		"parsePermissionsIntoHTML": func(perms []permission.Permission) template.HTML {
+			var output strings.Builder
+			for _, p := range perms {
+				output.WriteString("<tr>")
+				output.WriteString(fmt.Sprintf("<th>%d</th>", p.PermissionID))
+				output.WriteString(fmt.Sprintf("<td>%s</td>", p.Name))
+				output.WriteString(fmt.Sprintf("<td>%s</td>", p.Description))
+				output.WriteString(fmt.Sprintf("<td>%d</td>", p.Roles))
+				output.WriteString(fmt.Sprintf("<td><a href='/internal/permission/%d'>View</a></td>", p.PermissionID))
+				output.WriteString("</tr>")
+			}
+			return template.HTML(output.String())
+		},
+		"parsePermissionIntoHTML": func(p user.PermissionTemplate, perms []permission.Permission) template.HTML {
+			roleAdmin := t.permissionsParser(perms, permissions.ManageMembersGroup.String())
+			var output, roles strings.Builder
+			if len(p.Roles) > 0 {
+				roles.WriteString("Inherited by: <ol>")
+				for _, r := range p.Roles {
+					if roleAdmin {
+						roles.WriteString("<li style='list-style-type: none;'><span class='tab'></span>")
+						roles.WriteString(fmt.Sprintf("<a href='/internal/role/%d'>%s</a>", r.RoleID, r.Name))
+						roles.WriteString("</li>")
+					} else {
+						roles.WriteString("<li style='list-style-type: none;'><span class='tab'></span>")
+						roles.WriteString(fmt.Sprintf("%s", r.Name))
+						roles.WriteString("</li>")
+					}
+				}
+				roles.WriteString("</ol>")
+			}
+			output.WriteString("<p>")
+			output.WriteString(fmt.Sprintf("Permission ID: %d<br>", p.PermissionID))
+			output.WriteString(fmt.Sprintf("Name: %s<br>", p.Name))
+			output.WriteString(fmt.Sprintf("Description: %s<br><br>", p.Description))
+			output.WriteString(fmt.Sprintf("%s", roles.String()))
+			output.WriteString("</p>")
+			return template.HTML(output.String())
```

Comment:
I realised why this seemed off to me - part of the reason why `html/template` has a bit of a performance hit is that it automatically escapes everything to mitigate against XSS attacks. Wrapping this in a `template.HTML` is saying "I know this is safe, trust me bro", when it might not be (e.g. if `p.Description` contains a `<script>` we've just injected that into the HTML with no escaping).

At minimum, we should think about using `template.HTMLEscape` or its friends to ensure this is safe, though at that point the value over using `html/template` directly is questionable.

Reply:
That is true, I think it was slow is because it was doing the escape against every single iteration, I'll trial it before and after with doing it once and see if it is any faster or noticeably slower, whichever is more relevant

Reply:
I'm still of the opinion that we should move back to using `html/template` directly, and fix any performance hotspots that it exposes - it takes care of this for us without risking introducing a XSS vulnerability because we forgot to wrap one field somewhere in a `template.HTMLEscape` (especially since the code is public).

Reply:
I think so too after testing with users page, difference is negligible, previously it was written poorly by me

Reply:
I will rewrite it

Reply:
Rewritten